### DISCUSSION TITLE
Event loop closing when reusing an AsyncEngine

### DISCUSSION THREAD

**Discussion by abhay-apella** [sqlalchemy/sqlalchemy#11507] [Usage Questions]
I am attempting to build a shared singleton object that maintains the connection pool that can shared by my FastAPI webservers worker process thread pool. I created a health check function that executes a trivial query to ensure db connectivity by invoking `async_sessionmaker` and then executing the query and having the session context manager close. The health check function is called via a rest endpoint that is called once every 5s. However, on attempting to run this I frequently get event loop closed errors much like here:
https://github.com/sqlalchemy/sqlalchemy/discussions/5994

I notice the answer there is that I should dispose the engine every single time as well. However, I don't wish to do that as I want to maintain using a shared connection pool between my webserver threads. Is this a potential bug or am I doing something wrong?

Using: 
```
SQLAlchemy = {version = "2.0.28", extras = ["asyncio"]}
asyncpg = "^0.29.0"
```
--------------
Code
--------------

I build the engine instance as a callable singleton the returns the session maker and engine associated with it.
```
def __call__(...):
  ...
  self.engine_instance = create_async_engine(url=url, **all_args)
  sync_session_instance = sessionmaker(bind=self.engine_instance.sync_engine, expire_on_commit=False)
  async_session_instance = async_sessionmaker(
      bind=self.engine_instance, expire_on_commit=False, sync_session_class=sync_session_instance
  )
```

I define a health check function that is periodically called (once every 5s) that creates a session, executes a simple query and closes. 
```
async def health_check() -> None:
    _, session_maker = engine_singleton_instance()
    async with session_maker() as session:
        async with session.begin():
            try:
                await session.execute(text("SELECT 1"))
            except Exception as e:
                logging.error("Health Check Failed")
                raise e
```

I've also attempted to refactor the session_maker instance into a per thread call so that only the engine is created per process:
```
    engine = engine_singleton_instance()
    session_maker = get_session_maker(engine)
    async with session_maker() as session:
            async with session.begin():
                try:
                    await session.execute(text("SELECT 1"))
                except Exception as e:
                    logging.error("Health Check Failed")
                    raise e
```
This however still results in:
`RuntimeError: Event loop is closed`

**Comment by abhay-apella** [comment]
I think I understand the issue a bit better, it seems if I run `asyncio.run(health_check())` on app start up it does not close the connection and will only invalidate the connection on a subsequent call, after which the subsequent calls will succeed per worker process. This means I have to remove the `asyncio.run` and run it syncrhonously. Any reason why using `asyncio.run` would cause connection to be invalidated? It seems the context manager is properly exiting the connection. Is it that `asyncio.run` creates a separate event loop that causes issues?

**Reply by CaselIT** [reply]
> Is it that asyncio.run creates a separate event loop that causes issues?

yes, `run` creates a new event loop each time: https://docs.python.org/3/library/asyncio-runner.html#asyncio.run
